NRF52 serial: Fix UART console RX

naveenkaje opened the pull request:
While investigating the RX issue on NRF52_DK after SDK 14 updates,
it is observed that the RX FIFO doesn't get filled up, when the
flow control is disabled. Hence the readable never returns true.
If using Serial interface, the stdio file handles (0, 1, 2) get opened.
This results in configuring the flow control for STDIO, and it is observed
that the RX FIFO gets filled.

However, if RawSerial is used, the STDIO file handles
don't get opened. During the debug process it was observed that if the
flow control is configured once and then set to disabled, RX worked
as expected.

Alternative to this approach is that user application specifically
enables flow control as done in mbed's Greentea test suite. See https://goo.gl/r8nBYH

See https://goo.gl/8VB2qg step 14 for _initio's description.
See test code to reproduce the issue and test fix here: https://goo.gl/AQU1xG

### Description

The change in behavior with NRF52's UART RX is documented here. https://github.com/ARMmbed/mbed-os/issues/6891
This change is a fix for the above issue.

### Pull request type
    [ x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

naveenkaje commented:
This change is tested on NRF52_DK with SDK14 with GCC and ARM compilers.

naveenkaje commented:
This behavior is seen only on NRF52_DK (PCA10040). The test program works correctly on NRF528340_DK (PCA10056).

kjbracey commented:
This seems like a hack. We now have the mechanism for the target to specify that flow control should be used on the console UART and which pins (
https://github.com/ARMmbed/mbed-os/pull/6603).

Is this mechanism not working, or are you not using it?

kjbracey commented:
If the issue is that some programs are using `RawSerial` or `Serial` to access the console, without specifying flow control, then those programs are in error. This appears to be a good example of why programs shouldn't be doing that - it's not portable. 

It assumes that the console is on a serial port at all, and usually assumes no flow control. And probably assumes stuff about baud rate and format too.

To some extent you can fudge serial settings by picking up the same macros that mbed_retarget.cpp uses to specify console pins and settings, which is what Greentea is doing, but still it's bad - you're double-initialising the serial port if it is a serial port, and failing totally if it isn't.

Programs should be using C `stdout`/`stderr` or POSIX `FILENO_STDOUT`/`FILENO_STDERR` to access the console. I have an open issue for Greentea relating to that: https://github.com/ARMmbed/mbed-os/issues/6674, and I've just submitted a related PR for fault handling: #8076.

I'm not very keen on putting a hack into the HAL to ignore the settings it's told to use.

c1728p9 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/serial_api.c:
```diff
@@ -822,7 +822,14 @@ static void nordic_nrf5_uart_configure_object(serial_t *obj)
                                 uart_object->cts);
-    }
+    }  else { /* NRF_UART_HWFC_DISABLED */
+        /* Check if pin is set before configuring it */
```
Is this comment still relevant?

c1728p9 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/serial_api.c:
```diff
@@ -822,7 +822,14 @@ static void nordic_nrf5_uart_configure_object(serial_t *obj)
+        /* Check if pin is set before configuring it */
+			  uart_object->cts = NRF_UART_PSEL_DISCONNECTED;
+                          uart_object->rts = NRF_UART_PSEL_DISCONNECTED;
+			  nrf_uarte_hwfc_pins_set(nordic_nrf5_uart_register[uart_object->instance],
```
What is the purpose of this addition?

c1728p9 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/serial_api.c:
```diff
@@ -1082,6 +1089,19 @@ void serial_init(serial_t *obj, PinName tx, PinName rx)
+		 * Even though we set the flow control once here, it gets reset subsequentially.
+		 * This seems necessary until a better fix is found as users can specify RawSerial
+		 * interface with no flow control, resulting in broken RX.
+		 */
```
So by default  UART0 will use flow control connected to the terminal:
- Flow control enabled by default
- Flow control can be explicitly enabled or disabled by user by calling `Serial::set_flow_control`

If this is the case it seems reasonable to me.

c1728p9 commented:
> I'm not very keen on putting a hack into the HAL to ignore the settings it's told to use.

One possible alternative to changing the uart initialization is to on boot configure `STDIO_UART_CTS` as a gpio driven low for the nrf5x. That way DAPLink will see the NRF5X as ready to receive data when flow control is turned off. Enabling flow control explicitly (or via stdout or other use) will override the GPIO automatically.

kjbracey commented:
I was under a misconception here, that `Serial` et al were totally enforcing default format at the C++ layer - 9600 baud, 8N1, no flow control.

In fact they only default to 9600 baud, while the flow and format are left unmodified from however `serial_init` sets them.

Given that, having the HAL preset the flow mode for that UART so that use of `Serial` without an explicit `set_flow_control` does default to flow control isn't actively fighting against the higher layers _quite_ as much as I thought.

But the HAL does need to assume pins to do that, which is a cheat - activating 2 pins which weren't mentioned in the `serial_init`.

Maybe that's acceptable, but such magic should be more specific in the code - this isn't really a "NRF52" thing, it's specific to a board (isn't it?), and you want to activate it when configuring a UART against those particular pins you know are connected to the debug port on that particular board. There are a pile of board abstraction layers like `pin_instance_uart` underneath this function you're modifying, and you're effectively sidestepping them here. There needs to a bit more framework to it.

It feels like it would be a lot of complication to add magic so that code does what you think it means rather than what it says - making it so that

     Serial(A, B)

magically knows to also wiggle pins C and D without ever being told to do so because it makes that code work on a particular board.

I tend to agree with @marcuschangarm's abstraction comments on #6891. I'm not sure behind- the-scenes magic is the way to go here. Someone asking for `Serial(A,B)` is explicitly asking for 2-pin serial - if you are going to add 2 pins quietly, that could surprise folks, and even if it does make some insufficiently-abstracted programs work quietly, I can imagine people coming along later being confused about why it works - probably more confusing than why it currently doesn't work.

We already have mechanisms to abstract away from such an explicit serial request for the console. If you really need to explicitly reference the console by creating your own `Serial`-type object, we have console-specifying macros to configure the deeper settings as Greentea does. But ideally you shouldn't reference serial at all and just use stdin/stdout/STDIN_FILENO/STDOUT_FILENO, treating the console as device-independent.

(It may be we're still lacking a little in the abstraction area - I note you can find `STDIN_FILENO` which gives you most of the abilities via the POSIX API, but you can't get at the actual `FileHandle` object for the few things not available via the POSIX API, such as the sigio callback. I'm pondering exposing that)

kjbracey commented:
> One possible alternative to changing the uart initialization is to on boot configure STDIO_UART_CTS as a gpio driven low for the nrf5x. That way DAPLink will see the NRF5X as ready to receive data when flow control is turned off. Enabling flow control explicitly (or via stdout or other use) will override the GPIO automatically.

That's sounds like a good solution to me. Obviously it means you won't actually have working data flow control on a `Serial` if you don't specify it - you'll just get uncontrolled input and output - but then that's normal behaviour.

0xc0170 commented:
@kjbracey-arm @c1728p9 Should this PR be updated to the proposed solution (CTS driven low for nrf5x ) ?

Commit:
NRF52: setup the UART_RTS pin during startup to enable console RX

While investigating the RX issue on NRF52_DK after SDK 14 updates,
it is observed that the RX FIFO doesn't get filled up, when the
flow control is disabled. Hence the readable never returns true.
If using Serial interface, the stdio file handles (0, 1, 2) get opened.
This results in configuring the flow control for STDIO, and it is observed
that the RX FIFO gets filled.

However, if RawSerial is used, the STDIO file handles
don't get opened. During the debug process it was observed that if the
flow control is configured once and then set to disabled, RX worked
as expected.

Alternative to this approach is that user application specifically
enables flow control as done in mbed's Greentea test suite. See https://goo.gl/r8nBYH

See https://goo.gl/8VB2qg step 14 for _initio's description.
See test code to reproduce the issue and test fix here: https://goo.gl/AQU1xG

Description
The change in behavior with NRF52's UART RX is documented here. #6891
This change is a fix for the above issue.

Commit:
NRF52: setup the UART_RTS pin during startup to enable console RX

While investigating the RX issue on NRF52_DK after SDK 14 updates,
it is observed that the RX FIFO doesn't get filled up, when the
flow control is disabled. Hence the readable never returns true.
If using Serial interface, the stdio file handles (0, 1, 2) get opened.
This results in configuring the flow control for STDIO, and it is observed
that the RX FIFO gets filled.

However, if RawSerial is used, the STDIO file handles
don't get opened. During the debug process it was observed that if the
flow control is configured once and then set to disabled, RX worked
as expected.

Alternative to this approach is that user application specifically
enables flow control as done in mbed's Greentea test suite. See https://goo.gl/r8nBYH

See https://goo.gl/8VB2qg step 14 for _initio's description.
See test code to reproduce the issue and test fix here: https://goo.gl/AQU1xG

Description
The change in behavior with NRF52's UART RX is documented here. #6891
This change is a fix for the above issue.

naveenkaje commented:
Thanks @c1728p9 I made a modification to the fix in the latest patchset.
The issue is only seen on DAPLINK based boards. Here is the test results with this test code
https://goo.gl/AQU1xG

![capture](https://user-images.githubusercontent.com/39923816/46372051-a5bad100-c64f-11e8-9148-aa7df44ff2b9.PNG)

studavekar commented:
cc @ARMmbed/mbed-os-maintainers @marcuschangarm do we need this change as well for https://github.com/ARMmbed/mbed-os/pull/8292

naveenkaje commented:
This fix is independent of #8292

marcuschangarm commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/reloc_vector_table.c:
```diff
@@ -110,3 +112,11 @@ void nrf_reloc_vector_table(void)
+	gpio_t rts;
+	gpio_init_out(&rts, STDIO_UART_RTS);
+	/* Set STDIO_UART_RTS as gpio driven low */
+	gpio_write(&rts, 0);
```
This call will trigger an assert for boards with STDIO_UART_RTS set to NC. 

It would be better to wrap everything in an `#if` macro.

naveenkaje replied:
done

cmonr commented:
/morph build

cmonr commented:
IAR network license issue.
/morph mbed2-build
